### PR TITLE
Fix 2105

### LINKED ISSUE #2105: add group add / delete in bulk and group update in bulk
Hi,
With @mboudet we're working on https://github.com/genouest/ldapollo to synchronize our ldap users/groups with our Apollo configured with REMOTE_USER.
I'm testing it with our 500 users, and I sometimes see this error when assigning users to groups (both newly created):

```
Cannot invoke method getMetaData() on null object. Stacktrace follows:
java.lang.NullPointerException: Cannot invoke method getMetaData() on null object
	at org.bbop.apollo.GroupController.$tt__updateMembership(GroupController.groovy:414)
	at grails.plugin.cache.web.filter.PageFragmentCachingFilter.doFilter(PageFragmentCachingFilter.java:198)
	at grails.plugin.cache.web.filter.AbstractFilter.doFilter(AbstractFilter.java:63)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:449)
	at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
	at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
	at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
	at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:383)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
Do you see what could be the problem? Rerunning the script works, but usually fails on the next empty group.

### PR DESCRIPTION
**Pull request by nathandunn**
fixes #2105 

- [x] add group_update_membership
- ~~update update_memberships script~~
- [x] redo rest doc and review
- [x] refactor group add
- [x] add group_delete.groovy
- [x] refactor into GroupService
- [x] make GroupController delete work properly

### CODE REVIEW

**Comment by nathandunn** [comment]
{"membership": {<groupId>: ["user1", "user2", "user3"], <another-groupId>: ["user2", "user8"]}}

**Comment by nathandunn** [comment]
@abretaud  this is ready to test . . . 

Update_memberships I've tested the least, but the individual commands still work, so I imagine it still works.    Let me know.

Make sure you check the web services doc: 

```
Bulk memberships (instead of users and groupId) to update of the form: [ {groupId: <groupId>,users: ["user1", "user2", "user3"]}, {groupId:<another-groupId>, users: ["user2", "user8"]}]
```

File: grails-app/controllers/org/bbop/apollo/GroupController.groovy [hunk 1]
```diff
@@ -403,51 +387,26 @@ class GroupController {
...
-        // allow global admin, group creator, and group admin to update the group membership
-        if (!permissionService.hasGlobalPermissions(dataObject, GlobalPermissionEnum.ADMIN) && !(creatorMetaData && currentUser.id.toString() == creatorMetaData) && !permissionService.isGroupAdmin(groupInstance, currentUser)) {
 
-            render status: HttpStatus.UNAUTHORIZED.value()
-            return
-        }
-        log.info "Trying to update user group membership"
+        if(dataObject.memberships) {
 
+            JSONObject memberships = dataObject.memberships
```
Comment (abretaud) [inline grails-app/controllers/org/bbop/apollo/GroupController.groovy]: Tested! Group creation works fine.
For membership, I have this exception on this line:
```
Cannot cast object '[{"groupId":"1624","users":["..."]}, ...] with class 'org.codehaus.groovy.grails.web.json.JSONArray' to class 'org.codehaus.groovy.grails.web.json.JSONObject' due to: groovy.lang.GroovyRuntimeException: Could not find matching constructor for: org.codehaus.groovy.grails.web.json.JSONObject(org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject, org.codehaus.groovy.grails.web.json.JSONObject)	at org.bbop.apollo.GroupController.$tt__updateMembership(GroupController.groovy:401)
 	at grails.plugin.cache.web.filter.PageFragmentCachingFilter.doFilter(PageFragmentCachingFilter.java:198)
	at grails.plugin.cache.web.filter.AbstractFilter.doFilter(AbstractFilter.java:63)
 	at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:449)
         at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
         at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
         at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
         at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:383)
         at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
         at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
         at java.lang.Thread.run(Thread.java:748)
```

**Comment by nathandunn** [comment]
@abretaud  I had forgotten to add a push . . I also (hopefully) fixed the other error. 

Let me know if this fixes it or not.   Feel free to merge it directly if it does.

**Comment by abretaud** [comment]
Ok, tested everything, and it works, thanks!

**Pull request merged**